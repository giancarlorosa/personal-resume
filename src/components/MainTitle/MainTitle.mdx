import { Meta, Canvas, Controls } from '@storybook/blocks';
import * as MainTitleStories from './MainTitle.stories.jsx';

<Meta of={MainTitleStories} />

# Main Title Component

The **Main Title** component was developed to serve the `H1` tag for the pages. Its implementation is very simple and just asks for a **string** as parameter. Instead of using **props**, it was decided that using **children** could be more semantic and readable. As this component will always be te first element of any page, its alignm will be centered and it wil have a small underline element below the tag.

## Implementation

This component has a very simple implementation. Just inform a **string** as children.
**DO NOT USE NODE INSIDE CHILDREN PROP!**

<Canvas of={MainTitleStories.Implementation} />

<Controls of={MainTitleStories.Implementation} />

## Title followed by a paragraph

This is an implementation example using **Main Title** component followed by a **paragraph**. Here we can check component botton space.

<Canvas of={MainTitleStories.TitleWithParagraph} />

<Controls of={MainTitleStories.TitleWithParagraph} />
