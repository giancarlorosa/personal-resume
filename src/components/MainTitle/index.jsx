import PropTypes from 'prop-types';
import classNames from 'classnames';

function MainTitle({ children }) {
  const mainTitleClasses = classNames(
    'flex',
    'flex-col',
    'items-center',
    'font-bold',
    'mb-7',
    'text-4xl',
    'sm:text-5xl',
    'lg:text-6xl',
    'leading-tight',
    'sm:leading-tight',
    'lg:leading-tight',
    'text-dark-jungle-green',
    'text-center',
    'after:bg-mercury',
    'after:block',
    'after:content-[""]',
    'after:h-0.5',
    'after:mt-4',
    'after:w-14'
  );
  return <h1 className={mainTitleClasses}>{children}</h1>;
}

MainTitle.propTypes = {
  children: PropTypes.string.isRequired,
};

export default MainTitle;
