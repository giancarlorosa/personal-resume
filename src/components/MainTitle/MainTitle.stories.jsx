import MainTitle from '.';

export default {
  title: 'Components/Main Title',
  component: MainTitle,
};

export const Implementation = {
  args: {
    children: 'Main Title',
  },
};

export const OneLineBigTitle = {
  args: {
    children: 'Very big title text but without any linebreak',
  },
  render: ({ children }) => <MainTitle>{children}</MainTitle>,
};

export const MultipleLines = {
  args: {
    children: 'Very big title with multiple lines of text',
  },
  render: ({ children }) => (
    <div className="max-w-2xl mx-auto">
      <MainTitle>{children}</MainTitle>
    </div>
  ),
};

export const TitleWithParagraph = {
  args: {
    children: 'Title followed by a paragraph',
  },
  render: ({ children }) => (
    <div className="max-w-3xl mx-auto">
      <MainTitle>{children}</MainTitle>
      <p>
        Lorem ipsum dolor sit amet, consectetur adipisicing elit. Assumenda,
        odit harum minus autem dolore ab, hic magnam, sunt iste ipsam itaque
        architecto enim perferendis vel molestiae magni velit consequuntur
        veritatis.
      </p>
      <p>
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Magni rerum, id
        corporis ab, nesciunt asperiores hic doloribus voluptatem sunt quidem
        doloremque saepe delectus velit, cupiditate ut nam officia eum
        dignissimos.
      </p>
    </div>
  ),
};

export const MultipleLinesWithParagraph = {
  args: {
    children:
      'Very big title with multiple lines of text followed by a paragraph',
  },
  render: ({ children }) => (
    <div className="max-w-3xl mx-auto">
      <MainTitle>{children}</MainTitle>
      <p>
        Lorem ipsum dolor sit amet, consectetur adipisicing elit. Assumenda,
        odit harum minus autem dolore ab, hic magnam, sunt iste ipsam itaque
        architecto enim perferendis vel molestiae magni velit consequuntur
        veritatis.
      </p>
      <p>
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Magni rerum, id
        corporis ab, nesciunt asperiores hic doloribus voluptatem sunt quidem
        doloremque saepe delectus velit, cupiditate ut nam officia eum
        dignissimos.
      </p>
    </div>
  ),
};
