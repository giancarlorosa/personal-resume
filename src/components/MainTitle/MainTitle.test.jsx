/* eslint-disable no-console */
import { render } from '@testing-library/react';
import MainTitle from '.';

describe('MainTitle', () => {
  beforeAll(() => {
    jest.spyOn(console, 'error').mockImplementation(() => {});
  });

  afterAll(() => {
    console.error.mockRestore();
  });

  afterEach(() => {
    console.error.mockClear();
  });

  it('renders the component', () => {
    const { getByText } = render(<MainTitle>Main Title</MainTitle>);
    const componentTitle = getByText('Main Title');
    expect(componentTitle).toBeTruthy();
  });

  it('verifies if component has a H1 tag', () => {
    const { getByRole } = render(<MainTitle>Component text</MainTitle>);
    const componentTag = getByRole('heading', { level: 1 });
    expect(componentTag).toBeTruthy();
  });

  it('throws an error when not pass children prop', () => {
    render(<MainTitle />);

    expect(console.error).toHaveBeenCalled();
    expect(console.error.mock.calls[0][2]).toContain(
      'The prop `children` is marked as required in `MainTitle`'
    );
  });

  it('throws an error when children prop is not a string', () => {
    render(
      <MainTitle>
        <div>Main Title</div>
      </MainTitle>
    );

    expect(console.error).toHaveBeenCalled();
    expect(console.error.mock.calls[0][2]).toContain(
      'Invalid prop `children` of type `object` supplied to `MainTitle`'
    );
  });

  it('verifies changes in the component style', () => {
    const { container } = render(<MainTitle>Main Title</MainTitle>);

    expect(container.firstChild.className).toMatchSnapshot();
  });
});
