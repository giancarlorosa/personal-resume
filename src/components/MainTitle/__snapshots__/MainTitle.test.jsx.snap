// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainTitle verifies changes in the component style 1`] = `"flex flex-col items-center font-bold mb-7 text-4xl sm:text-5xl lg:text-6xl leading-tight sm:leading-tight lg:leading-tight text-dark-jungle-green text-center after:bg-mercury after:block after:content-[""] after:h-0.5 after:mt-4 after:w-14"`;
